cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -445,38 +442,183 @@ index<T, IdxT> build(
+      if (curr_query_size == final_graph_size) {
+        curr_topk         = topk;
+        curr_itopk_size   = topk * 2;
+        curr_graph_degree = graph_degree;
+      }
+
+      cuvs::neighbors::cagra::search_params search_params;
+      search_params.algo        = cuvs::neighbors::cagra::search_algo::AUTO;
+      search_params.max_queries = max_chunk_size;
+      search_params.itopk_size  = curr_itopk_size;
+
+      // Create an index (idx), a query view (dev_query_view), and a mdarray for
+      // search results (neighbors).
+      auto dev_dataset_view = raft::make_device_matrix_view<const T, int64_t>(
+        dev_dataset.data_handle(), (int64_t)curr_graph_size, dev_dataset.extent(1));
+
+      auto idx = index<T, IdxT>(
+        res, params.metric, dev_dataset_view, raft::make_const_mdspan(cagra_graph.view()));
+
+      auto dev_query_view = raft::make_device_matrix_view<const T, int64_t>(
+        dev_dataset.data_handle(), (int64_t)curr_query_size, dev_dataset.extent(1));
+      auto neighbors = raft::make_host_matrix<IdxT, int64_t>(curr_query_size, curr_topk);
+
+      // Search.
+      // Since there are many queries, divide them into batches and search them.
+      cuvs::spatial::knn::detail::utils::batch_load_iterator<T> query_batch(
+        dev_query_view.data_handle(),
+        curr_query_size,
+        dev_query_view.extent(1),
+        max_chunk_size,
+        raft::resource::get_cuda_stream(res),
+        raft::resource::get_workspace_resource(res));
+      for (const auto& batch : query_batch) {
+        auto batch_dev_query_view = raft::make_device_matrix_view<const T, int64_t>(
+          batch.data(), batch.size(), dev_query_view.extent(1));
+        auto batch_dev_neighbors_view = raft::make_device_matrix_view<IdxT, int64_t>(
+          dev_neighbors.data_handle(), batch.size(), curr_topk);
+        auto batch_dev_distances_view = raft::make_device_matrix_view<float, int64_t>(
+          dev_distances.data_handle(), batch.size(), curr_topk);
+
+        cuvs::neighbors::cagra::search(res,
+                                       search_params,
+                                       idx,
+                                       batch_dev_query_view,
+                                       batch_dev_neighbors_view,
+                                       batch_dev_distances_view);
+
+        auto batch_neighbors_view = raft::make_host_matrix_view<IdxT, int64_t>(
+          neighbors.data_handle() + batch.offset() * curr_topk, batch.size(), curr_topk);
+        raft::copy(batch_neighbors_view.data_handle(),
+                   batch_dev_neighbors_view.data_handle(),
+                   batch_neighbors_view.size(),
+                   raft::resource::get_cuda_stream(res));
+      }
+
+      // Optimize graph
+      bool flag_last  = (curr_graph_size == final_graph_size);
+      curr_graph_size = curr_query_size;
+      cagra_graph     = raft::make_host_matrix<IdxT, int64_t>(curr_graph_size, curr_graph_degree);
```

Comment:
There are cases when we work with limited host mem, therefore it might make sense to delete the existing graph before we allocate the new one. 
```suggestion
cagra_graph = raft::make_host_matrix<IdxT, int64_t>(0, 0); // delete existing graph
cagra_graph = raft::make_host_matrix<IdxT, int64_t>(curr_graph_size, curr_graph_degree);
```
Alternatively, it might be even better to allocate the final graph in advance before the while loop, and here just construct a view for the actual graph size. This way host OOM could be caught earlier in the process.

Reply:
Thanks, will fix in that way.